libbeat/outputs/shipper/shipper_test.go

```diff
@@ -93,6 +102,236 @@ func TestConvertMapStr(t *testing.T) {
+						EventId: "first",
+					},
+					{
+						EventId: "second",
+					},
+				},
+			},
+			events: events,
+			exp:    []publisher.Event{},
+		},
+		{
+			name: "partially accepted events return retries",
+			resp: &sc.PublishReply{
+				Results: []*sc.EventResult{
+					{
+						EventId: "second",
+					},
+				},
+			},
+			events: events,
+			exp:    events[:1],
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			retries := findRetries(tc.resp, tc.events)
+			require.Equal(t, tc.exp, retries)
+		})
+	}
+}
+
+func TestPublish(t *testing.T) {
+	events := []beat.Event{
+		{
+			Timestamp: time.Now(),
+			Meta:      mapstr.M{"event": "first"},
+			Fields:    mapstr.M{"foo": "bar"},
+		},
+		{
+			Timestamp: time.Now(),
+			Meta:      mapstr.M{"event": "second"},
+			Fields:    mapstr.M{"a": "b"},
+		},
+	}
+
+	cases := []struct {
+		name       string
+		events     []beat.Event
+		expSignals []outest.BatchSignal
+		qSize      int
+	}{
+		{
+			name:   "sends a batch",
+			events: events[:1],
+			expSignals: []outest.BatchSignal{
+				{
+					Tag: outest.BatchACK,
+				},
+			},
+			qSize: 5,
+		},
+		{
+			name:   "retries not accepted events",
+			events: events,
+			expSignals: []outest.BatchSignal{
+				{
+					Tag:    outest.BatchRetryEvents,
+					Events: toPublisherEvents(events[1:]),
+				},
+			},
+			qSize: 1,
+		},
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+			defer cancel()
+			producer := sc.NewProducerMock(tc.qSize)
+			grpcServer := grpc.NewServer()
+			sc.RegisterProducerServer(grpcServer, producer)
+
+			listener, err := net.Listen("tcp", "localhost:0") // random available port
+			require.NoError(t, err)
+
+			defer grpcServer.Stop()
+			go func() {
+				_ = grpcServer.Serve(listener)
+			}()
+
+			cfg, err := config.NewConfigFrom(map[string]interface{}{
+				"server": listener.Addr().String(),
+			})
+			require.NoError(t, err)
+
+			group, err := makeShipper(
+				nil,
+				beat.Info{Beat: "libbeat", IndexPrefix: "testbeat"},
+				outputs.NewNilObserver(),
+				cfg,
+			)
+			require.NoError(t, err)
+			require.Len(t, group.Clients, 1)
+
+			batch := outest.NewBatch(tc.events...)
+			err = group.Clients[0].Publish(ctx, batch)
+			require.NoError(t, err)
+
+			require.Equal(t, tc.expSignals, batch.Signals)
+		})
+	}
+
+	t.Run("cancel the batch when the server is not available", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+		defer cancel()
+
+		producer := sc.NewProducerMock(5)
+		grpcServer := grpc.NewServer()
+		sc.RegisterProducerServer(grpcServer, producer)
+
+		listener, err := net.Listen("tcp", "localhost:0") // random available port
+		require.NoError(t, err)
+
+		defer grpcServer.Stop()
+		// run with a 2 second delay
+		go func() {
+			<-time.After(2 * time.Second)
```

Comment:
+1 I would prefer if we can get rid of this fixed delay

Reply:
We just need to run it in timeout tolerances (after the first publishing attempt failed with a timeout). There is no really a workaround that I could find. Feel free to propose something better.

Reply:
I tried changing this myself as an exercise to understand the code. You do need some delay or polling between starting the gRPC server and making the first request. The server doesn't start instantly. I attached the changes as a patch (with a .txt extension because that's all GitHub accepts...).

[patch.txt](https://github.com/elastic/beats/files/8663851/patch.txt)

Reply:
@cmacknz I like your idea to use `require.Eventually` in this case. I've applied your patch in https://github.com/elastic/beats/pull/31558/commits/4177a4286b4a33378566a76fecc45d0b4a3c38b2